homeassistant/components/switcher_kis/config_flow.py

```diff
@@ -2,9 +2,97 @@
 
 from __future__ import annotations
 
-from homeassistant.helpers import config_entry_flow
+from collections.abc import Mapping
+from typing import Any
 
-from .const import DOMAIN
-from .utils import async_has_devices
+import voluptuous as vol
 
-config_entry_flow.register_discovery_flow(DOMAIN, "Switcher", async_has_devices)
+from homeassistant.config_entries import ConfigEntry, ConfigFlow, ConfigFlowResult
+
+from .const import CONF_TOKEN, CONF_USERNAME, DATA_DISCOVERY, DOMAIN
+from .utils import async_discover_devices
+
+
+class SwitcherFlowHandler(ConfigFlow, domain=DOMAIN):
+    """Handle Switcher config flow."""
+
+    VERSION = 1
+
+    entry: ConfigEntry | None = None
+    username: str | None = None
+    token: str | None = None
+
+    async def async_step_user(
+        self, user_input: dict[str, Any] | None = None
+    ) -> ConfigFlowResult:
+        """Handle the start of the config flow."""
+        if self._async_current_entries(True):
+            return self.async_abort(reason="single_instance_allowed")
+
+        self.hass.data.setdefault(DOMAIN, {})
+        if DATA_DISCOVERY not in self.hass.data[DOMAIN]:
+            self.hass.data[DOMAIN][DATA_DISCOVERY] = self.hass.async_create_task(
+                async_discover_devices()
+            )
+
+        return self.async_show_form(step_id="confirm")
+
+    async def async_step_confirm(
+        self, user_input: dict[str, Any] | None = None
+    ) -> ConfigFlowResult:
+        """Handle user-confirmation of the config flow."""
+        discovered_devices = await self.hass.data[DOMAIN][DATA_DISCOVERY]
+
+        if len(discovered_devices) == 0:
+            self.hass.data[DOMAIN].pop(DATA_DISCOVERY)
+            return self.async_abort(reason="no_devices_found")
+
+        return self.async_create_entry(
+            title="Switcher",
+            data={
+                CONF_USERNAME: self.username,
+                CONF_TOKEN: self.token,
+            },
```

Comment:
Please first check if one of the discovered devices needs a token before asking the user to add a token

Reply:
I do that here:
https://github.com/home-assistant/core/pull/123578/files#diff-3e4eacfcf78f613ccf0f87e8409a1e040758f7e7ec9bcb802becd3de9458b96cR57-R61
I am calling reauth to get username and token only if device that needs a token found

Reply:
This needs to happen in 2 places:
1. Same place you already added, this will handle devices that are detected during runtime that needs a token.
2. When user add the integration, if a device that needs a token is detect you need to switch to asking for credentials before continuing to create an entry.

Exampe:
https://github.com/home-assistant/core/blob/21987a67e7c4d3c526ddfbd4866c64f93c2569f0/homeassistant/components/shelly/config_flow.py#L171-L172

Reply:
> This needs to happen in 2 places:
> 
> 1. Same place you already added, this will handle devices that are detected during runtime that needs a token.
> 2. When user add the integration, if a device that needs a token is detect you need to switch to asking for credentials before continuing to create an entry.
> 
> Exampe:
> 
> https://github.com/home-assistant/core/blob/21987a67e7c4d3c526ddfbd4866c64f93c2569f0/homeassistant/components/shelly/config_flow.py#L171-L172

Maybe I am missing something but when I tested this part I got the "reauth" when I initially set the integration for the first time and when I had this already set up and I pressed reload on the integration (And when I didn't have the device there and restart HA)..

Plus this is the only part where the bridge is detecting the devices so I can determine if this is a token-based device no?

Reply:
I put the missing step and did a little refactoring. Please revisit.